File: Minimum Spanning Tree - GFG/minimum-spanning-tree.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends

class DSU{
   
   private:
   vector<int>Parent;
   vector<int>Rank;
   
   public:
   DSU(int n){
       Parent.resize(n);
       Rank.resize(n);
       for(int i=0;i<n;i++){
           Parent[i] = i;
           Rank[i] = 1;
       }
   }
   int find(int x){
       if(Parent[x] == x) return x;
       return Parent[x] = find(Parent[x]);
   }
   bool Union(int x,int y){
       int lx = find(x);
       int ly = find(y);
       if(lx != ly){
           if(Rank[lx] > Rank[ly]){
               Parent[ly] = lx;
           }
           else{
               Parent[lx] = ly;
               if(Rank[lx] == Rank[ly]){
                   Rank[lx]++;
               }
           }
           return true;
       }
       else{
           return false;
       }
   }
   
};

class Solution
{
	public:
	//Function to find sum of weights of edges of the Minimum Spanning Tree.
    int spanningTree(int n, vector<vector<int>> graph[])
    {
        vector<vector<int>>edges;
        for(int node=0;node<n;node++){
            for(auto &nbrs : graph[node]){
                int nbr = nbrs[0];
                int edgeWt = nbrs[1];
                edges.push_back({edgeWt,node,nbr});
            }
        }
        sort(edges.begin(),edges.end());
        DSU obj(n);
        int sum = 0;
        for(int i=0;i<edges.size();i++){
            int edgeWt = edges[i][0];
            int u = edges[i][1];
            int v = edges[i][2];
            if(obj.Union(u,v)){
                sum += edgeWt;
            }
        }
        return sum;
        
    }
};

//{ Driver Code Starts.


int main()
{
    int t;
    cin >> t;
    while (t--) {
        int V, E;
        cin >> V >> E;
        vector<vector<int>> adj[V];
        int i=0;
        while (i++<E) {
            int u, v, w;
            cin >> u >> v >> w;
            vector<int> t1,t2;
            t1.push_back(v);
            t1.push_back(w);
            adj[u].push_back(t1);
            t2.push_back(u);
            t2.push_back(w);
            adj[v].push_back(t2);
        }
        
        Solution obj;
    	cout << obj.spanningTree(V, adj) << "\n";
    }

    return 0;
}


// } Driver Code Ends

File: Minimum Spanning Tree - GFG/minimum-spanning-tree.java
//{ Driver Code Starts


import java.util.*;
import java.io.*;
import java.lang.*;

public class Main{
	static BufferedReader br;
	static PrintWriter ot;
    public static void main(String args[]) throws IOException {
		br = new BufferedReader(new InputStreamReader(System.in));
		ot = new PrintWriter(System.out);
		int t = Integer.parseInt(br.readLine().trim());
		while(t-- > 0){
			String s[] = br.readLine().trim().split(" ");
			int V = Integer.parseInt(s[0]);
			int E = Integer.parseInt(s[1]);
			int edges[][] = new int[E][3];
			for(int i = 0; i < E; i++){
				s = br.readLine().trim().split(" ");
				edges[i][0] = Integer.parseInt(s[0]);
				edges[i][1] = Integer.parseInt(s[1]);
				edges[i][2] = Integer.parseInt(s[2]);
			}
			ot.println(new Solution().spanningTree(V, E, edges));
		}
		ot.close();
	}
}
// } Driver Code Ends


// User function Template for Java

class Pair implements Comparable<Pair>{
    int node;
    int wt;
    Pair(int node, int wt){
        this.node = node;
        this.wt = wt;
    }
    public int compareTo(Pair p2){
        return this.wt - p2.wt;
    }
}

class Solution{
	static int spanningTree(int n, int m, int edges[][]){
	    int sumMST = 0;
	    ArrayList<Pair> graph[] = new ArrayList[n];
	    for(int i=0;i<n;i++){
	        graph[i] = new ArrayList<Pair>();
	    }
	    PriorityQueue<Pair> pq = new PriorityQueue<>();
	    boolean visited[] = new boolean[n];
	    pq.add(new Pair(0,0));
	    for(int i=0;i<m;i++){
	        int u = edges[i][0];
	        int v = edges[i][1];
	        int wt = edges[i][2];
	        graph[u].add(new Pair(v,wt));
	        graph[v].add(new Pair(u,wt));
	    }
	    while(!pq.isEmpty()){
	        int node = pq.peek().node;
	        int wt = pq.peek().wt;
	        pq.remove();
	        if(visited[node] == true) continue;
	        visited[node] = true;
	        sumMST += wt;
	        for(Pair p : graph[node]){
	            int nbr = p.node;
	            int edgeWt = p.wt;
	            if(visited[nbr] == true) continue;
	            pq.add(new Pair(nbr,edgeWt));
	        }
	    }
	    return sumMST;
	}
}